beacon_node/beacon_chain/tests/store_tests.rs

```diff
@@ -445,40 +498,93 @@ fn delete_blocks_and_states() {
     }
 
     // After all that, the chain dump should still be OK
-    check_chain_dump(&harness, unforked_blocks + fork_blocks + 1);
+    check_chain_dump(&harness, unforked_blocks + fork_blocks - 1);
 }
 
 // Check that we never produce invalid blocks when there is deep forking that changes the shuffling.
 // See https://github.com/sigp/lighthouse/issues/845
 fn multi_epoch_fork_valid_blocks_test(
     initial_blocks: usize,
-    num_fork1_blocks: usize,
-    num_fork2_blocks: usize,
+    num_fork1_blocks_: usize,
+    num_fork2_blocks_: usize,
     num_fork1_validators: usize,
 ) -> (TempDir, TestHarness, Hash256, Hash256) {
     let db_path = tempdir().unwrap();
     let store = get_store(&db_path);
-    let harness = get_harness(store.clone(), LOW_VALIDATOR_COUNT);
+    let validators_keypairs =
+        types::test_utils::generate_deterministic_keypairs(LOW_VALIDATOR_COUNT);
+    let mut harness =
+        BeaconChainHarness::new_with_disk_store(MinimalEthSpec, store, validators_keypairs);
+
+    let mut num_fork1_blocks: u64 = num_fork1_blocks_.try_into().unwrap();
+    let mut num_fork2_blocks: u64 = num_fork2_blocks_.try_into().unwrap();
 
     // Create the initial portion of the chain
     if initial_blocks > 0 {
-        harness.extend_chain(
-            initial_blocks,
-            BlockStrategy::OnCanonicalHead,
-            AttestationStrategy::AllValidators,
-        );
+        let initial_slots: Vec<Slot> = (1..=initial_blocks).map(Into::into).collect();
+        let state = harness.get_current_state();
+        let all_validators = harness.get_all_validators();
+        harness.add_attested_blocks_at_slots(state, &initial_slots, &all_validators);
     }
 
+    let mut head1 = harness.chain.head().unwrap().beacon_block_root;
+    let mut head2 = head1;
+
     assert!(num_fork1_validators <= LOW_VALIDATOR_COUNT);
     let fork1_validators: Vec<usize> = (0..num_fork1_validators).collect();
     let fork2_validators: Vec<usize> = (num_fork1_validators..LOW_VALIDATOR_COUNT).collect();
 
-    let (head1, head2) = harness.generate_two_forks_by_skipping_a_block(
-        &fork1_validators,
-        &fork2_validators,
-        num_fork1_blocks,
-        num_fork2_blocks,
-    );
+    let mut first_fork_epoch = true;
+
+    let mut slot = harness.get_current_slot() + 1;
+    let mut fork1_state = harness.get_current_state();
+    let mut fork2_state = fork1_state.clone();
+    while num_fork1_blocks > 0 || num_fork2_blocks > 0 {
```

Comment:
Similarly I think this loop could be expressed in terms of `add_blocks_on_multiple_chains`?

Reply:
Done ✅